### PR TITLE
Rework the fog falloff function

### PR DESCRIPTION
**Pull request by rreusser**
***Edit: I'm rewriting this description to reflect updated findings.***

This PR reworks the fog function to satisfy some basic constraints. Fog:

1. does not need to be physically accurate
2. should decay exponentially(ish)
3. onset at the near limit should be at least C2 continuous (smooth second derivative) to avoid a visible band
4. opacity should be 0% at the near limit
5. opacity should be 100% at the far limit
6. Dark fog and light fog should obscure the terrain similarly

To satisfy 1-4, I've used the function:

Screen Shot 2021-03-28 at 12 00 06 PM

with n=near, f=far, p and q arbitrary constants. With some additional clipping to precisely satisfy 5, we get:

Screen Shot 2021-03-28 at 12 00 52 PM

with c chosen slightly greater than 1.0 to force fog to 1.0 at the far limit. This function looks like:
https://www.desmos.com/calculator/zxu1xd3zjf
Screen Shot 2021-03-28 at 11 57 01 AM

The sixth criteria gave me a lot of trouble. This is because I mistakenly thought there was something "correct" about gamma-correct blending, not appreciating that linear RGB blending is only concerned with dealing with amounts of light on a linear scale so that you can correctly add and subtract them. I've since realized that, no matter how many blog posts people write about gradients in sRGB space being evil, I think sRGB is a better fit for this case since the fog is not physically accurate and since we want dark and light fog with identical near and far limits to obscure the terrain similarly.

In the end, I threw out gamma correction and reverted it to a simple `mix()`.

### Mixing in sRGB (good)
Light and dark fog over empty water now look roughly the same: 

Screen Shot 2021-03-28 at 12 08 41 PMScreen Shot 2021-03-28 at 12 09 02 PM

### Linear RGB without raising to the third power

(Apologies, I lost the parameters so the limits are slightly different than above, though what's illustrated here is mainly the banding) Without the third power which softens the onset, we only have C0 continuity and what in my opinion is a really unpleasant band at the onset of the fog. I think the third power achieves C3 continuity or something? Looks plenty smooth anyway.

Screen Shot 2021-03-28 at 12 36 17 PMScreen Shot 2021-03-28 at 12 36 02 PM

### Mixing in linear RGB (bad)
Fog mixed in linear RGB looks like this. It's very simply flipped in terms of the amount of light emitted, but that looks very different to our eyes, which perceive on a logarithmic scale.

Screen Shot 2021-03-28 at 12 11 32 PMScreen Shot 2021-03-28 at 12 11 22 PM

### Mixing in sRGB (good)
With terrain and heavy fog, mixing in sRGB space results in fog which obscures the terrain very similarly:

nogamma-lightnogamma-dark

### Mixing in linear RGB (bad)
Mixing in linear RGB space results in a very different amount of fog, which is entirely what I was trying to avoid in the first place by thinking I needed to worry about gamma, and which is now somewhat obvious.

gamma-lightgamma-dark

### Mixing in sRGB (good)
Things seem perfectly reasonable and consistent to control with satellite imagery as well.

Screen Shot 2021-03-28 at 12 26 04 PMScreen Shot 2021-03-28 at 12 25 54 PM

### Mixing in linear RGB (bad)

Linear RGB doesn't look bad at all, but it makes the near and far limits have a very different effect. More important than colors blended perfectly in linear space (for what reason, exactly, since there's already little physical basis?), I think, is that once you configure the amount of fog, you shouldn't be surprised by its appearance on light or dark maps.

Screen Shot 2021-03-28 at 12 32 06 PMScreen Shot 2021-03-28 at 12 32 37 PM

In conclusion, I think this achieves a controllable look free of artifacts. I think there is a place for gamma correction, but this is not it.

### CODE REVIEW

**Comment by rreusser** [comment]
Raising the clipped exponential to a power is another option I might try again:

https://www.desmos.com/calculator/1f49cct0ri

Screen Shot 2021-03-26 at 2 52 32 PM

File: src/shaders/_prelude_fog.fragment.glsl [hunk 1]
```diff
@@ -22,25 +22,19 @@ float fog_opacity(vec3 position) {
...
+    // Fog power puts the fog at about 0.6% and 99.4% at near and far, respectively
+    // The multiplier puts the far limit at 1.0 since clipping depends on that
+    const float fog_pow = 10.0;
+    float fog_falloff = min(1.0, 1.00675 / (1.0 + exp(-fog_pow * ((depth - start) / (end - start) - 0.5))));
+
+    return fog_falloff * u_fog_opacity;
 }
 
 vec3 fog_apply(vec3 color, vec3 position) {
-    return gamma_mix(
```
Comment (karimnaaji) [inline src/shaders/_prelude_fog.fragment.glsl:39]: Before we go ahead with removing, I wanted to be sure that the screenshots are a fair comparison. If I were to singularize the bands and sample right at the limits, tops and bottoms don't match (in the whites).

Right tends to full white srgb(1,1,1) while left tends to srgb(0.8,0.8,0.8). If we were to use the same colors, would we still have that illusion of reducing the range?

Screen Shot 2021-03-26 at 3 47 16 PM

**Review by karimnaaji — APPROVED** [review]
Looks good, thanks for the nice description comparisons and details!
